### ISSUE TITLE
//undo does not work

### ISSUE THREAD

**Issue by lilian-gnt** [SBPrime/AsyncWorldEdit-Premium#157]
I had a complete building here, I // cut then // undo unfortunately there is more thing of my building : https://puu.sh/Cq56V/c7d6ee322f.png

AsyncWorldEdit - Premium 3.6.0-rc-03
worldedit-bukkit-7.0.0-beta-04.jar
worldguard-legacy-7.0.0-beta-02.jar
Spigot : git-Spigot-29ab5e4-2bc7d1d (MC: 1.13.2) (Implementing API version 1.13.2-R0.1-SNAPSHOT)

**Comment by lilian-gnt** [comment]
```
[12:38:20 INFO]: TotoArc issued server command: //undo
[12:38:21 INFO]: [AsyncWorldEdit] ***********************************
[12:38:21 INFO]: [AsyncWorldEdit] Unable to create undo iterator for ts.1680e5a8386.87a50056.0
[12:38:21 INFO]: [AsyncWorldEdit] ***********************************
[12:38:21 INFO]: [AsyncWorldEdit] *
[12:38:21 INFO]: [AsyncWorldEdit] * Exception: java.io.FileNotFoundException
[12:38:21 INFO]: [AsyncWorldEdit] * Error message: plugins/AsyncWorldEdit/undo/03fdd38a-6526-4529-8e0b-54f8beb18f34/ts.1680e5a8386.87a50056.0 (No such file or directory)
[12:38:21 INFO]: [AsyncWorldEdit] * Stack: 
[12:38:21 INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.open0(Native Method)
[12:38:21 INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.open(RandomAccessFile.java:316)
[12:38:21 INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.<init>(RandomAccessFile.java:243)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.changesetSerializer.iterators.BaseFileIterator.initializeStream(BaseFileIterator.java:121)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.changesetSerializer.iterators.UndoFileBackwordIterator.initializeStream(UndoFileBackwordIterator.java:88)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSet.initializeIterator(FileChangeSet.java:325)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSet.backwardIterator(FileChangeSet.java:134)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.ThreadSafeChangeSet.backwardIterator(ThreadSafeChangeSet.java:100)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.MemoryMonitorChangeSet.backwardIterator(MemoryMonitorChangeSet.java:171)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.ThreadSafeEditSession.doUndo(ThreadSafeEditSession.java:806)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.UndoProcessor.processUndo(UndoProcessor.java:78)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.CancelabeEditSession.doUndo(CancelabeEditSession.java:372)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.CancelabeEditSession.undo(CancelabeEditSession.java:368)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.AsyncEditSession$1.task(AsyncEditSession.java:155)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.AsyncTask.doRun(AsyncTask.java:71)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.BaseTask.run(BaseTask.java:152)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils.processTaskQueue(SchedulerUtils.java:168)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils.access$000(SchedulerUtils.java:62)
[12:38:21 INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils$1.run(SchedulerUtils.java:131)
[12:38:21 INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftTask.run(CraftTask.java:81)
[12:38:21 INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52)
[12:38:21 INFO]: [AsyncWorldEdit] * java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[12:38:21 INFO]: [AsyncWorldEdit] * java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[12:38:21 INFO]: [AsyncWorldEdit] * java.lang.Thread.run(Thread.java:748)
[12:38:21 INFO]: [AsyncWorldEdit] ***********************************
```

**Comment by SBPrime** [comment]
Are you able to provide a log for the operation that you tried to undo? This looks like a problem with entites.

**Comment by lilian-gnt** [comment]
```
[12:13:17] [Server thread/INFO]: HookDonn_ lost connection: Timed out
[12:13:17] [Server thread/INFO]:  | HookDonn_ a quitt? le serveur !
[12:35:46] [Server thread/INFO]: TotoArc issued server command: /gmc
[12:37:19] [Server thread/INFO]: TotoArc issued server command: //copy
[12:37:22] [Server thread/INFO]: TotoArc issued server command: /back
[12:37:30] [Server thread/INFO]: TotoArc issued server command: /gmc
[12:37:39] [Server thread/INFO]: TotoArc issued server command: //paste
[12:37:47] [Server thread/INFO]: TotoArc issued server command: //undo
[12:37:56] [User Authenticator #14/INFO]: UUID of player HookDonn_ is 3f1c313e-d8c9-48b7-b12c-44ab5703a82a
[12:37:57] [Server thread/INFO]: HookDonn_[/??.???.???.???:?????] logged in with entity id ?????? at ([MineDream]1362.0098823487376, 57.50342554526373, 3304.480808685351)
[12:37:57] [Server thread/INFO]:  | HookDonn_ a rejoint le serveur !
[12:38:16] [Server thread/INFO]: HookDonn_ issued server command: /gmc
[12:38:20] [Server thread/INFO]: TotoArc issued server command: //undo
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] ***********************************
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] Unable to create undo iterator for ts.1680e5a8386.87a50056.0
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] ***********************************
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] *
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * Exception: java.io.FileNotFoundException
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * Error message: plugins/AsyncWorldEdit/undo/03fdd38a-6526-4529-8e0b-54f8beb18f34/ts.1680e5a8386.87a50056.0 (No such file or directory)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * Stack: 
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.open0(Native Method)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.open(RandomAccessFile.java:316)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.io.RandomAccessFile.<init>(RandomAccessFile.java:243)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.changesetSerializer.iterators.BaseFileIterator.initializeStream(BaseFileIterator.java:121)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.changesetSerializer.iterators.UndoFileBackwordIterator.initializeStream(UndoFileBackwordIterator.java:88)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSet.initializeIterator(FileChangeSet.java:325)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSet.backwardIterator(FileChangeSet.java:134)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.ThreadSafeChangeSet.backwardIterator(ThreadSafeChangeSet.java:100)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.history.changeset.MemoryMonitorChangeSet.backwardIterator(MemoryMonitorChangeSet.java:171)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.ThreadSafeEditSession.doUndo(ThreadSafeEditSession.java:806)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.UndoProcessor.processUndo(UndoProcessor.java:78)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.CancelabeEditSession.doUndo(CancelabeEditSession.java:372)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.CancelabeEditSession.undo(CancelabeEditSession.java:368)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.AsyncEditSession$1.task(AsyncEditSession.java:155)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.AsyncTask.doRun(AsyncTask.java:71)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.worldedit.BaseTask.run(BaseTask.java:152)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils.processTaskQueue(SchedulerUtils.java:168)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils.access$000(SchedulerUtils.java:62)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.primesoft.asyncworldedit.utils.SchedulerUtils$1.run(SchedulerUtils.java:131)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftTask.run(CraftTask.java:81)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * org.bukkit.craftbukkit.v1_13_R2.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] * java.lang.Thread.run(Thread.java:748)
[12:38:21] [Craft Scheduler Thread - 2226/INFO]: [AsyncWorldEdit] ***********************************
```

I have nothing more in the logs. :/
On the first //undo there is no error, but from the first it did not work.

**Comment by PlasmaPod** [comment]
Im having same issue so decided to check the worledit forums and found this http://forum.enginehub.org/threads/bug-history-does-not-get-saved.26607/

**Comment by PlasmaPod** [comment]
_**[EDIT]** Undo Works on WorldEdit 7.0.0 Beta 4
https://dev.bukkit.org/projects/worldedit/files/2646337
but breaks other plugins that i just updated for the new version xD that rely on a dependancy of the new WE dev update_ 
_**[EDIT]2** need to downgrade to this version of worldguard also  7.0.0 Beta 2
https://dev.bukkit.org/projects/worldguard/files/2642132_

Worldedit Version : 7.0.0-SNAPSHOT;3997-ba8bde0/plugi
Spigot Version : git-spigot-19c3c5a-5932f8a (MC: 1.13.2) (latest build version of spigot at time of writing this)

Plugins
`Plugins (62): WhoWas, PlaceholderAPI, Arconix, SuperTrails, DiscordInvite, NoteBlockAPI, TreeGravity, CoreProtect, dynmap, Scuba, EggCatcher, OpenInv, Essentials, Elevator, WorldBorder, Seashells, ActionHealth, MeltDown, ActionBarAPI, MoreRecipes, ProtocolLib, Multiverse-Core, LibsDisguises, NoCheatPlus, ImageMaps, OmNom, AsyncWorldEdit, PermissionsEx, NoWeather, mcMMO, SwagChat, Vault, ColoredAnvils, SimpleSit, ClearChat, BKCommonLib, SignLink, EpicRename, EpicHeads, LWC, HolographicDisplays, EssentialsAntiBuild, EssentialsSpawn, BungeeTabListPlus, PlayerHeads, WorldEdit, SynthCrates, Towny, BannerMaker, Train_Carts, WorldGuard, Quests, TCHangRail, EditableSign, TownyNameUpdater, Dynmap-Towny, VentureChat, AreaShop, Jobs, SilkSpawners, ChestShop, ArmorStandTools`

CoreProtect Config
`# Logs changes made via the plugin "WorldEdit" if it's in use on your server.
worldedit: false`

Log
`[01:53:14] [Server thread/INFO]: PlasmaPod issued server command: //cut
[01:53:14] [Thread-34/WARN]: Exception in thread "Thread-34" 
[01:53:14] [Thread-34/WARN]: java.lang.NoSuchMethodError: com.sk89q.worldedit.history.change.BlockChange.getCurrent()Lcom/sk89q/worldedit/world/block/BlockStateHolder;
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.changesetSerializer.serializers.SerializerBlockChange.serialize(SerializerBlockChange.java:98)
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.changesetSerializer.SerializerManager.serialize(SerializerManager.java:502)
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.changesetSerializer.SerializerManager.save(SerializerManager.java:473)
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.changesetSerializer.SerializerManager.save(SerializerManager.java:313)
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSet.save(FileChangeSet.java:210)
[01:53:14] [Thread-34/WARN]: 	at org.primesoft.asyncworldedit.worldedit.history.changeset.FileChangeSetManager.saveLoop(FileChangeSetManager.java:126)
[01:53:14] [Thread-34/WARN]: 	at java.lang.Thread.run(Thread.java:748)`

**Comment by SBPrime** [comment]
Looks like you are using a dev version of WorldEdit. I do not support dev builds, I support only beta releases. Dev builds support might be added in near future.

**Comment by PlasmaPod** [comment]
> Looks like you are using a dev version of WorldEdit. I do not support dev builds, I support only beta releases. Dev builds support might be added in near future.

Its been working Perfectly now that ive found the correct build of worldedit to use glad moved to this over Fastasyncworldedit and i love the block byb block animation as all the blocks come pasting in that is so freaking awesome

**Comment by SBPrime** [comment]
@PlasmaPod 
| i love the block byb block animation as all the blocks come pasting in that is so freaking awesome
Ye a feature that was added simply for debug, ended up being extremely fun to watch (and for youtube).

So the undo problem fixed?

**Comment by PlasmaPod** [comment]
yeah it was just the dev builds of worldedit and worldguard that broke it it might be worth mentioning on the spigot page to use worldedit stable builds as instead of a hard to find commit ID for a specific build

**Comment by veteranmina** [comment]
Issue found, change was made from BlockStateHolder to BaseBlock in multiple areas with this commit: https://github.com/EngineHub/WorldEdit/commit/3fefcbf97183cc2cf8fff36baf8d76908c5891f3#diff-1ad0bc21372d0844746d419cfaa897e9
Currently AWE is calling for BlockStateHolder:
https://github.com/SBPrime/AsyncWorldEdit/blob/a2b6e20b933958249ce4054aebb594814d6ad343/AsyncWorldEdit/src/main/java/org/primesoft/asyncworldedit/changesetSerializer/serializers/SerializerBlockChange.java
https://github.com/SBPrime/AsyncWorldEdit/blob/a2b6e20b933958249ce4054aebb594814d6ad343/AsyncWorldEdit/src/main/java/org/primesoft/asyncworldedit/changesetSerializer/serializers/SerializerBiomeChange.java
https://github.com/SBPrime/AsyncWorldEdit/blob/a2b6e20b933958249ce4054aebb594814d6ad343/AsyncWorldEdit/src/main/java/org/primesoft/asyncworldedit/worldedit/history/change/BiomeChange.java

**Comment by SBPrime** [comment]
@veteranmina Unfortunately this is not the case. The error is caused by something else because it happens with compatible versions of WE. The change that you send causes incompatibility with dev build (with is not supported).

**Comment by SBPrime** [comment]
Undo fixed in 3.6.0
